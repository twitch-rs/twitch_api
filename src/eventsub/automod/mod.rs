//! Automod related events

use super::{EventSubscription, EventType};
use serde_derive::{Deserialize, Serialize};

pub mod message;
pub mod settings;
pub mod terms;

#[doc(inline)]
pub use message::{AutomodMessageHoldV1, AutomodMessageHoldV1Payload};
#[doc(inline)]
pub use message::{AutomodMessageHoldV2, AutomodMessageHoldV2Payload};
#[doc(inline)]
pub use message::{AutomodMessageUpdateV1, AutomodMessageUpdateV1Payload};
#[doc(inline)]
pub use message::{AutomodMessageUpdateV2, AutomodMessageUpdateV2Payload};

#[doc(inline)]
pub use terms::{AutomodTermsUpdateV1, AutomodTermsUpdateV1Payload};

#[doc(inline)]
pub use settings::{AutomodSettingsUpdateV1, AutomodSettingsUpdateV1Payload};

/// A category identified by automod for a message.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[non_exhaustive]
#[serde(rename_all = "snake_case")]
pub enum AutomodCategory {
    /// Aggressive behavior or language.
    Aggressive,
    /// Bullying or harassment.
    Bullying,
    /// Discrimination or negative references to disabilities.
    Disability,
    /// Content relating to or expressing sexuality.
    Sexuality,
    /// Inappropriate or suggestive sexual terms
    Sexwords,
    /// Misogynistic language or behavior.
    Misogyny,
    /// Racist language or behavior.
    Racism,
    /// Profane language.
    Profanity,
    /// Use of swear words or expletives.
    Swearing,
    /// An unknown category identified by Automod, contains the raw string provided by Twitch.
    #[serde(untagged)]
    Unknown(String),
}
