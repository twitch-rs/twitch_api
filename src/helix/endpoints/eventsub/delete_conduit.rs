//! Deletes a specified conduit.
//! [`delete-conduit`](https://dev.twitch.tv/docs/api/reference/#delete-conduit)
//!
//! Note that it may take some time for Eventsub subscriptions on a deleted conduit to show as disabled when calling [Get Eventsub Subscriptions](super::get_eventsub_subscriptions).

use super::*;
use helix::RequestDelete;

/// Query Parameters for [Delete Conduit](super::delete_conduit)
///
/// [`delete-conduit`](https://dev.twitch.tv/docs/api/reference/#delete-conduit)
#[derive(PartialEq, Eq, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[must_use]
#[non_exhaustive]
pub struct DeleteConduitRequest<'a> {
    /// Conduit ID.
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    #[cfg_attr(feature = "deser_borrow", serde(borrow = "'a"))]
    pub id: Cow<'a, types::ConduitIdRef>,
}

impl<'a> DeleteConduitRequest<'a> {
    /// Delete a specific conduit
    pub fn new(id: impl types::IntoCow<'a, types::ConduitIdRef> + 'a) -> Self {
        Self { id: id.into_cow() }
    }
}

/// Return Values for [Delete Conduit](super::delete_conduit)
///
/// [`delete-conduit`](https://dev.twitch.tv/docs/api/reference/#delete-conduit)
#[derive(PartialEq, Eq, Deserialize, Serialize, Debug, Clone)]
#[non_exhaustive]
pub enum DeleteConduitResponse {
    /// 204 - Conduit deleted
    Success,
}

impl Request for DeleteConduitRequest<'_> {
    type Response = DeleteConduitResponse;

    const PATH: &'static str = "eventsub/conduits";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = twitch_oauth2::validator![];
}

impl RequestDelete for DeleteConduitRequest<'_> {
    fn parse_inner_response(
        request: Option<Self>,
        uri: &http::Uri,
        response: &str,
        status: http::StatusCode,
    ) -> Result<helix::Response<Self, Self::Response>, helix::HelixRequestDeleteError>
    where
        Self: Sized,
    {
        match status {
            http::StatusCode::NO_CONTENT | http::StatusCode::OK => Ok(helix::Response::with_data(
                DeleteConduitResponse::Success,
                request,
            )),
            _ => Err(helix::HelixRequestDeleteError::InvalidResponse {
                reason: "unexpected status",
                response: response.to_string(),
                status,
                uri: uri.clone(),
            }),
        }
    }
}

#[cfg(test)]
#[test]
fn test_response() {
    helix::assert_helix_snapshot!(
        DeleteConduitRequest:
        req = DeleteConduitRequest::new("bfcfc993-26b1-b876-44d9-afe75a379dac"),
        @r#"
    uri
    ----
    https://api.twitch.tv/helix/eventsub/conduits?id=bfcfc993-26b1-b876-44d9-afe75a379dac

    response
    ----
    Response {
        data: Success,
        pagination: None,
        request: Some(
            DeleteConduitRequest {
                id: "bfcfc993-26b1-b876-44d9-afe75a379dac",
            },
        ),
        total: None,
        other: None,
    }
    "#,
    );
}
