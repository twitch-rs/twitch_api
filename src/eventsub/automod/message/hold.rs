#![doc(alias = "automod.message.hold")]
//! A message was caught by automod for review

use super::*;
/// [`automod.message.hold`](https://dev.twitch.tv/docs/eventsub/eventsub-subscription-types/#automodmessagehold): a message was caught by automod for review.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
#[non_exhaustive]
pub struct AutomodMessageHoldV1 {
    /// User ID of the broadcaster (channel).
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    pub broadcaster_user_id: types::UserId,
    /// User ID of the moderator.
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    pub moderator_user_id: types::UserId,
}

impl AutomodMessageHoldV1 {
    /// Get automod hold notifications for this channel as a moderator
    pub fn new(
        broadcaster_user_id: impl Into<types::UserId>,
        moderator_user_id: impl Into<types::UserId>,
    ) -> Self {
        Self {
            broadcaster_user_id: broadcaster_user_id.into(),
            moderator_user_id: moderator_user_id.into(),
        }
    }
}

impl EventSubscription for AutomodMessageHoldV1 {
    type Payload = AutomodMessageHoldV1Payload;

    const EVENT_TYPE: EventType = EventType::AutomodMessageHold;
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator =
        twitch_oauth2::validator![twitch_oauth2::Scope::ModeratorManageAutoMod];
    const VERSION: &'static str = "1";
}

/// [`automod.message.hold`](AutomodMessageHoldV1) response payload.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
#[non_exhaustive]
pub struct AutomodMessageHoldV1Payload {
    /// The ID of the broadcaster specified in the request.
    pub broadcaster_user_id: types::UserId,
    /// The login of the broadcaster specified in the request.
    pub broadcaster_user_login: types::UserName,
    /// The user name of the broadcaster specified in the request.
    pub broadcaster_user_name: types::DisplayName,
    /// The message senderâ€™s user ID.
    pub user_id: types::UserId,
    /// The message senderâ€™s login name.
    pub user_login: types::UserName,
    /// The message senderâ€™s display name.
    pub user_name: types::DisplayName,
    /// The ID of the message that was flagged by automod.
    pub message_id: types::MsgId,
    /// The body of the message.
    pub message: AutomodMessage,
    /// The category of the message.
    pub category: super::AutomodCategory,
    /// The level of severity. Measured between 1 to 4.
    pub level: u8,
    /// The timestamp of when automod saved the message.
    pub held_at: types::Timestamp,
}

#[cfg(test)]
#[test]
fn parse_payload_v1() {
    use crate::eventsub::{Event, Message};

    let payload = r##"
    {
        "subscription": {
            "id": "e523fda0-01b6-4b0e-9024-a5a80c5ad680",
            "status": "enabled",
            "type": "automod.message.hold",
            "version": "1",
            "condition": {
                "broadcaster_user_id": "129546453",
                "moderator_user_id": "129546453"
            },
            "transport": {
                "method": "websocket",
                "session_id": "AgoQRniKAQ1ITYSESh4ku8anEBIGY2VsbC1j"
            },
            "created_at": "2024-10-19T20:11:13.917500523Z",
            "cost": 0
        },
        "event": {
            "broadcaster_user_id": "129546453",
            "broadcaster_user_login": "nerixyz",
            "broadcaster_user_name": "nerixyz",
            "user_id": "489584266",
            "user_login": "uint128",
            "user_name": "uint128",
            "message_id": "332e99ac-e19c-4368-a15b-793e8266b51f",
            "message": {
                "text": "boobs",
                "fragments": [
                    {
                        "type": "text",
                        "text": "boobs",
                        "cheermote": null,
                        "emote": null
                    }
                ]
            },
            "category": "sexwords",
            "level": 4,
            "held_at": "2024-10-19T20:11:16.799750627Z"
        }
    }
    "##;

    let val = Event::parse(payload).unwrap();
    crate::tests::roundtrip(&val);

    let Event::AutomodMessageHoldV1(val) = val else {
        panic!("invalid event type");
    };
    let Message::Notification(notif) = val.message else {
        panic!("invalid message type");
    };

    assert_eq!(notif.broadcaster_user_id.as_str(), "129546453");
    assert_eq!(notif.category, AutomodCategory::Sexwords);
    assert_eq!(notif.level, 4);
    assert_eq!(notif.message.fragments.len(), 1);
}

/// [`automod.message.hold`](https://dev.twitch.tv/docs/eventsub/eventsub-subscription-types/#automodmessagehold-v2): a message was caught by automod for review.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
#[non_exhaustive]
pub struct AutomodMessageHoldV2 {
    /// User ID of the broadcaster (channel).
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    pub broadcaster_user_id: types::UserId,
    /// User ID of the moderator.
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    pub moderator_user_id: types::UserId,
}

impl AutomodMessageHoldV2 {
    /// Get automod hold notifications for this channel as a moderator
    pub fn new(
        broadcaster_user_id: impl Into<types::UserId>,
        moderator_user_id: impl Into<types::UserId>,
    ) -> Self {
        Self {
            broadcaster_user_id: broadcaster_user_id.into(),
            moderator_user_id: moderator_user_id.into(),
        }
    }
}

impl EventSubscription for AutomodMessageHoldV2 {
    type Payload = AutomodMessageHoldV2Payload;

    const EVENT_TYPE: EventType = EventType::AutomodMessageHold;
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator =
        twitch_oauth2::validator![twitch_oauth2::Scope::ModeratorManageAutoMod];
    const VERSION: &'static str = "2";
}

/// [`automod.message.hold`](AutomodMessageHoldV2) response payload.
// XXX: this struct can't be deny-unknown-fields because of the flattened reason
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[non_exhaustive]
pub struct AutomodMessageHoldV2Payload {
    /// The ID of the broadcaster specified in the request.
    pub broadcaster_user_id: types::UserId,
    /// The login of the broadcaster specified in the request.
    pub broadcaster_user_login: types::UserName,
    /// The user name of the broadcaster specified in the request.
    pub broadcaster_user_name: types::DisplayName,
    /// The message senderâ€™s user ID.
    pub user_id: types::UserId,
    /// The message senderâ€™s login name.
    pub user_login: types::UserName,
    /// The message senderâ€™s display name.
    pub user_name: types::DisplayName,
    /// The ID of the message that was flagged by automod.
    pub message_id: types::MsgId,
    /// The body of the message.
    pub message: AutomodMessage,
    /// The reason why a message was held
    #[serde(flatten)]
    pub reason: AutomodHeldReason,
    /// The timestamp of when automod saved the message.
    pub held_at: types::Timestamp,
}

#[test]
fn parse_payload_v2_automod() {
    use crate::eventsub::{Event, Message};

    let payload = r##"
    {
        "subscription": {
            "id": "85c8dcb0-7af4-4581-b684-32087d386384",
            "status": "enabled",
            "type": "automod.message.hold",
            "version": "2",
            "condition": {
                "broadcaster_user_id": "129546453",
                "moderator_user_id": "129546453"
            },
            "transport": {
                "method": "websocket",
                "session_id": "AgoQREw4FYBWQ5quz4J-S4VYkRIGY2VsbC1j"
            },
            "created_at": "2024-11-18T16:36:08.691979783Z",
            "cost": 0
        },
        "event": {
            "broadcaster_user_id": "129546453",
            "broadcaster_user_login": "nerixyz",
            "broadcaster_user_name": "nerixyz",
            "user_id": "489584266",
            "user_login": "uint128",
            "user_name": "uint128",
            "message_id": "78ccd959-3e7e-4f8d-bd8b-f92c359b0a7d",
            "message": {
                "text": "ðŸ˜‚ ass",
                "fragments": [
                    {
                        "type": "text",
                        "text": "ðŸ˜‚ ",
                        "cheermote": null,
                        "emote": null
                    },
                    {
                        "type": "text",
                        "text": "ass",
                        "cheermote": null,
                        "emote": null
                    }
                ]
            },
            "reason": "automod",
            "automod": {
                "category": "swearing",
                "level": 4,
                "boundaries": [
                    {
                        "start_pos": 2,
                        "end_pos": 4
                    }
                ]
            },
            "blocked_term": null,
            "held_at": "2024-11-18T16:59:46.323937273Z"
        }
    }
    "##;

    let val = Event::parse(payload).unwrap();
    crate::tests::roundtrip(&val);

    let Event::AutomodMessageHoldV2(val) = val else {
        panic!("invalid event type");
    };
    let Message::Notification(notif) = val.message else {
        panic!("invalid message type");
    };

    assert_eq!(notif.broadcaster_user_id.as_str(), "129546453");
    assert_eq!(notif.message.fragments.len(), 2);

    let AutomodHeldReason::Automod { automod } = &notif.reason else {
        panic!("invalid held reason");
    };
    assert_eq!(automod.category, AutomodCategory::Swearing);
    assert_eq!(automod.level, 4);
    assert_eq!(
        automod.boundaries,
        &[AutomodMessageBoundary {
            start_pos: 2,
            end_pos: 4
        }]
    );
}

#[test]
fn parse_payload_v2_blocked_term() {
    use crate::eventsub::{Event, Message};

    let payload = r##"
    {
        "subscription": {
            "id": "85c8dcb0-7af4-4581-b684-32087d386384",
            "status": "enabled",
            "type": "automod.message.hold",
            "version": "2",
            "condition": {
                "broadcaster_user_id": "129546453",
                "moderator_user_id": "129546453"
            },
            "transport": {
                "method": "websocket",
                "session_id": "AgoQREw4FYBWQ5quz4J-S4VYkRIGY2VsbC1j"
            },
            "created_at": "2024-11-18T16:36:08.691979783Z",
            "cost": 0
        },
        "event": {
            "broadcaster_user_id": "129546453",
            "broadcaster_user_login": "nerixyz",
            "broadcaster_user_name": "nerixyz",
            "user_id": "489584266",
            "user_login": "uint128",
            "user_name": "uint128",
            "message_id": "dcfc6b48-0fd1-446c-8cf5-d1810bb55b73",
            "message": {
                "text": "boobs Kappa ðŸ˜‚ðŸ˜‚ foo private",
                "fragments": [
                    {
                        "type": "text",
                        "text": "boobs",
                        "cheermote": null,
                        "emote": null
                    },
                    {
                        "type": "text",
                        "text": " ",
                        "cheermote": null,
                        "emote": null
                    },
                    {
                        "type": "emote",
                        "text": "Kappa",
                        "cheermote": null,
                        "emote": {
                            "id": "25",
                            "emote_set_id": "0"
                        }
                    },
                    {
                        "type": "text",
                        "text": " ðŸ˜‚ðŸ˜‚ ",
                        "cheermote": null,
                        "emote": null
                    },
                    {
                        "type": "text",
                        "text": "foo",
                        "cheermote": null,
                        "emote": null
                    },
                    {
                        "type": "text",
                        "text": " ",
                        "cheermote": null,
                        "emote": null
                    },
                    {
                        "type": "text",
                        "text": "private",
                        "cheermote": null,
                        "emote": null
                    }
                ]
            },
            "reason": "blocked_term",
            "automod": null,
            "blocked_term": {
                "terms_found": [
                    {
                        "term_id": "e4d4f1ba-99bf-4b19-9875-cd4eda98ead9",
                        "owner_broadcaster_user_id": "129546453",
                        "owner_broadcaster_user_login": "nerixyz",
                        "owner_broadcaster_user_name": "nerixyz",
                        "boundary": {
                            "start_pos": 15,
                            "end_pos": 17
                        }
                    },
                    {
                        "term_id": "e60a94ea-e5d9-444e-a114-4cfd2f86c6ad",
                        "owner_broadcaster_user_id": "129546453",
                        "owner_broadcaster_user_login": "nerixyz",
                        "owner_broadcaster_user_name": "nerixyz",
                        "boundary": {
                            "start_pos": 19,
                            "end_pos": 25
                        }
                    }
                ]
            },
            "held_at": "2024-11-18T16:58:41.476117057Z"
        }
    }
    "##;

    let val = Event::parse(payload).unwrap();
    crate::tests::roundtrip(&val);

    let Event::AutomodMessageHoldV2(val) = val else {
        panic!("invalid event type");
    };
    let Message::Notification(notif) = val.message else {
        panic!("invalid message type");
    };

    assert_eq!(notif.broadcaster_user_id.as_str(), "129546453");
    assert_eq!(notif.message.fragments.len(), 7);

    let AutomodHeldReason::BlockedTerm { blocked_term } = &notif.reason else {
        panic!("invalid held reason");
    };
    assert_eq!(blocked_term.terms_found.len(), 2);
    assert_eq!(
        blocked_term.terms_found[0].term_id.as_str(),
        "e4d4f1ba-99bf-4b19-9875-cd4eda98ead9"
    );
    assert_eq!(
        blocked_term.terms_found[0].boundary,
        AutomodMessageBoundary {
            start_pos: 15,
            end_pos: 17
        }
    );
}
