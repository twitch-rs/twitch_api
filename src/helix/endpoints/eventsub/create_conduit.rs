//! Creates a new conduit for your Client.
//! [`create-conduit`](https://dev.twitch.tv/docs/api/reference/#create-conduits)

use super::*;
use crate::eventsub;
use helix::RequestPost;

/// Query Parameters for [Create Conduit](super::create_conduit)
///
/// [`create-conduit`](https://dev.twitch.tv/docs/api/reference/#create-conduits)
#[derive(PartialEq, Eq, Serialize, Clone, Debug, Default)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[must_use]
#[non_exhaustive]
pub struct CreateConduitRequest {}

impl Request for CreateConduitRequest {
    type Response = eventsub::Conduit;

    const PATH: &'static str = "eventsub/conduits";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = twitch_oauth2::validator![];
}

/// Body Parameters for [Create Conduit](super::create_conduit)
///
/// [`create-conduit`](https://dev.twitch.tv/docs/api/reference/#create-conduits)
#[derive(PartialEq, Eq, Deserialize, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[non_exhaustive]
pub struct CreateConduitBody {
    /// The number of shards to create for this conduit.
    pub shard_count: usize,
}

impl CreateConduitBody {
    /// Conduit body settings
    pub const fn new(shard_count: usize) -> Self { Self { shard_count } }
}

impl helix::private::SealedSerialize for CreateConduitBody {}

impl RequestPost for CreateConduitRequest {
    type Body = CreateConduitBody;

    fn parse_inner_response(
        request: Option<Self>,
        uri: &http::Uri,
        response: &str,
        status: http::StatusCode,
    ) -> Result<helix::Response<Self, Self::Response>, helix::HelixRequestPostError>
    where
        Self: Sized,
    {
        #[derive(PartialEq, Deserialize, Debug)]
        #[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
        struct InnerResponse {
            data: [eventsub::Conduit; 1],
        }

        let inner_response: InnerResponse = helix::parse_json(response, true).map_err(|e| {
            helix::HelixRequestPostError::DeserializeError(
                response.to_string(),
                e,
                uri.clone(),
                status,
            )
        })?;

        let [conduit] = inner_response.data;

        Ok(helix::Response::new(conduit, None, request, None, None))
    }
}

#[cfg(test)]
#[test]
fn test_response() {
    helix::assert_helix_snapshot!(
        CreateConduitRequest:
        req = CreateConduitRequest::default(),
        res = br#"{
            "data": [
                {
                    "id": "bfcfc993-26b1-b876-44d9-afe75a379dac",
                    "shard_count": 5
                }
            ]
        }"#,
    );
}

#[cfg(test)]
#[test]
fn test_successful_unexpected_response() {
    use helix::*;
    let req: CreateConduitRequest = CreateConduitRequest::default();

    let data = br#"{
      "data": []
    }
    "#
    .to_vec();
    let http_response = http::Response::builder().status(200).body(data).unwrap();

    let uri = req.get_uri().unwrap();
    let response = CreateConduitRequest::parse_response(Some(req), &uri, http_response);
    assert!(response.is_err());
}
