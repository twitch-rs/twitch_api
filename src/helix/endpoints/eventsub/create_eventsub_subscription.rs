//! Creates an EventSub subscription.
//!
//! See also [`HelixClient::create_eventsub_subscription`](crate::helix::HelixClient::create_eventsub_subscription)
use super::*;
use crate::eventsub::{EventSubscription, EventType, Status, Transport, TransportResponse};

/// Query Parameters for [Create EventSub Subscription](super::create_eventsub_subscription)
///
/// [`create-eventsub-subscription`](https://dev.twitch.tv/docs/api/reference#create-eventsub-subscription)
#[derive(PartialEq, Eq, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[non_exhaustive]
#[must_use]
pub struct CreateEventSubSubscriptionRequest<E: EventSubscription> {
    #[cfg_attr(feature = "typed-builder", builder(setter(skip), default))]
    #[serde(skip)]
    phantom: std::marker::PhantomData<E>,
}

impl<E: EventSubscription> CreateEventSubSubscriptionRequest<E> {
    /// Create a new eventsub subscription
    pub fn new() -> Self { Self::default() }
}

impl<E: EventSubscription> Default for CreateEventSubSubscriptionRequest<E> {
    fn default() -> Self {
        Self {
            phantom: Default::default(),
        }
    }
}

impl<E: EventSubscription> helix::Request for CreateEventSubSubscriptionRequest<E> {
    type Response = CreateEventSubSubscription<E>;

    #[cfg(feature = "twitch_oauth2")]
    const OPT_SCOPE: &'static [twitch_oauth2::Scope] = E::OPT_SCOPE;
    const PATH: &'static str = "eventsub/subscriptions";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = E::SCOPE;
}

/// Body Parameters for [Create EventSub Subscription](super::create_eventsub_subscription)
///
/// [`create-eventsub-subscription`](https://dev.twitch.tv/docs/api/reference#create-eventsub-subscription)
///
/// # Notes
///
/// This body is quite different from the official body. If you want the true representation in text, see [`helix::HelixRequestBody::try_to_body`] on [`CreateEventSubSubscriptionRequest<E: EventSubscription>`](CreateEventSubSubscriptionRequest)
#[derive(PartialEq, Eq, Deserialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[non_exhaustive]
pub struct CreateEventSubSubscriptionBody<E: EventSubscription> {
    /// Subscription that will be created
    #[serde(bound(deserialize = "E: EventSubscription"))]
    pub subscription: E,
    /// The notification delivery specific information
    pub transport: Transport,
}

impl<E: EventSubscription> helix::HelixRequestBody for CreateEventSubSubscriptionBody<E> {
    fn try_to_body(&self) -> Result<hyper::body::Bytes, helix::BodyError> {
        #[derive(PartialEq, Serialize, Debug)]
        struct IEventSubRequestBody<'a> {
            r#type: EventType,
            version: &'static str,
            condition: serde_json::Value,
            transport: &'a Transport,
        }

        let b = IEventSubRequestBody {
            r#type: E::EVENT_TYPE,
            version: E::VERSION,
            condition: self.subscription.condition()?,
            transport: &self.transport,
        };
        serde_json::to_vec(&b).map_err(Into::into).map(Into::into)
    }
}

// FIXME: Builder?
impl<E: EventSubscription> CreateEventSubSubscriptionBody<E> {
    /// Create a new [`CreateEventSubSubscriptionBody`]
    pub const fn new(subscription: E, transport: Transport) -> Self {
        Self {
            subscription,
            transport,
        }
    }
}

/// Return Values for [Create EventSub Subscription](super::create_eventsub_subscription)
///
/// [`create-eventsub-subscription`](https://dev.twitch.tv/docs/api/reference#create-eventsub-subscription)
#[derive(PartialEq, Eq, Deserialize, Serialize, Debug, Clone)]
#[non_exhaustive]
pub struct CreateEventSubSubscription<E: EventSubscription> {
    /// ID of the subscription created.
    pub id: types::EventSubId,
    /// Status of the subscription.
    pub status: Status,
    /// The category of the subscription that was created.
    #[serde(rename = "type")]
    pub type_: EventType,
    /// The version of the subscription type that was created.
    pub version: String,
    /// JSON object specifying custom parameters for the subscription.
    #[serde(bound(deserialize = "E: EventSubscription"))]
    pub condition: E,
    /// RFC3339 timestamp indicating when the subscription was created.
    pub created_at: types::Timestamp,
    /// JSON object indicating the notification delivery specific information. Includes the transport method and callback URL.
    pub transport: TransportResponse,
    /// Total number of subscriptions for the client ID that made the subscription creation request.
    pub total: usize,
    /// Total cost of all the subscriptions for the client ID that made the subscription creation request.
    pub total_cost: usize,
    /// The maximum total cost allowed for all of the subscriptions for the client ID that made the subscription creation request.
    pub max_total_cost: usize,
    /// How much the subscription counts against your limit.
    pub cost: usize,
}

impl<E: EventSubscription> helix::RequestPost for CreateEventSubSubscriptionRequest<E> {
    type Body = CreateEventSubSubscriptionBody<E>;

    fn parse_inner_response(
        request: Option<Self>,
        uri: &http::Uri,
        text: &str,
        status: http::StatusCode,
    ) -> Result<helix::Response<Self, Self::Response>, helix::HelixRequestPostError>
    where
        Self: Sized,
    {
        #[derive(PartialEq, Eq, Deserialize, Debug)]
        #[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
        pub struct InnerResponseData<E: EventSubscription> {
            cost: usize,
            #[serde(bound(deserialize = "E: EventSubscription"))]
            condition: E,
            created_at: types::Timestamp,
            id: types::EventSubId,
            status: Status,
            transport: TransportResponse,
            #[serde(rename = "type")]
            type_: EventType,
            version: String,
        }
        #[derive(PartialEq, Deserialize, Debug)]
        #[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
        struct InnerResponse<E: EventSubscription> {
            #[serde(bound(deserialize = "E: EventSubscription"))]
            data: Vec<InnerResponseData<E>>,
            limit: Option<usize>,
            total: usize,
            total_cost: usize,
            max_total_cost: usize,
        }
        let response: InnerResponse<E> = helix::parse_json(text, true).map_err(|e| {
            helix::HelixRequestPostError::DeserializeError(text.to_string(), e, uri.clone(), status)
        })?;
        let data = response.data.into_iter().next().ok_or_else(|| {
            helix::HelixRequestPostError::InvalidResponse {
                reason: "missing response data",
                response: text.to_string(),
                status,
                uri: uri.clone(),
            }
        })?;
        Ok(helix::Response::with_data(
            CreateEventSubSubscription {
                // helix::Response total is generally the total number of results, not what the total for this endpoint means. Thus, we set it to None.))
                total: response.total,
                total_cost: response.total_cost,
                max_total_cost: response.max_total_cost,
                cost: data.cost,
                id: data.id,
                status: data.status,
                type_: data.type_,
                version: data.version,
                condition: data.condition,
                created_at: data.created_at,
                transport: data.transport,
            },
            request,
        ))
    }
}

#[cfg(test)]
#[test]
fn test_request() {
    use crate::eventsub::{self, user::UserUpdateV1};

    let sub = UserUpdateV1::new("1234");
    let transport =
        eventsub::Transport::webhook("https://this-is-a-callback.com", "s3cre7".to_string());

    helix::assert_helix_snapshot!(CreateEventSubSubscriptionRequest::<UserUpdateV1>:
        req = CreateEventSubSubscriptionRequest::<UserUpdateV1>::default(),
        body = CreateEventSubSubscriptionBody::new(sub, transport),
        res = br#"{
            "data": [
                {
                    "id": "26b1c993-bfcf-44d9-b876-379dacafe75a",
                    "status": "webhook_callback_verification_pending",
                    "type": "user.update",
                    "version": "1",
                    "condition": {
                        "user_id": "1234"
                    },
                    "created_at": "2020-11-10T14:32:18.730260295Z",
                    "transport": {
                        "method": "webhook",
                        "callback": "https://this-is-a-callback.com"
                    },
                    "cost": 1
                }
            ],
            "total": 1,
            "total_cost": 1,
            "max_total_cost": 10000
        }"#,
        status = 202,
        @r#"
    uri
    ----
    https://api.twitch.tv/helix/eventsub/subscriptions?

    body
    ----
    {"type":"user.update","version":"1","condition":{"user_id":"1234"},"transport":{"method":"webhook","callback":"https://this-is-a-callback.com","secret":"s3cre7"}}

    response
    ----
    Response {
        data: CreateEventSubSubscription {
            id: "26b1c993-bfcf-44d9-b876-379dacafe75a",
            status: WebhookCallbackVerificationPending,
            type_: UserUpdate,
            version: "1",
            condition: UserUpdateV1 {
                user_id: "1234",
            },
            created_at: "2020-11-10T14:32:18.730260295Z",
            transport: Webhook(
                WebhookTransportResponse {
                    callback: "https://this-is-a-callback.com",
                },
            ),
            total: 1,
            total_cost: 1,
            max_total_cost: 10000,
            cost: 1,
        },
        pagination: None,
        request: Some(
            CreateEventSubSubscriptionRequest {
                phantom: PhantomData<twitch_api::eventsub::user::update::UserUpdateV1>,
            },
        ),
        total: None,
        other: None,
    }
    "#,
    );
}
