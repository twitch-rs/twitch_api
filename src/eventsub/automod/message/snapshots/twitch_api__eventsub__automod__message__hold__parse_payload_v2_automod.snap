---
source: src/eventsub/automod/message/hold.rs
expression: parsed
---
AutomodMessageHoldV2(
    Payload {
        subscription: EventSubscriptionInformation {
            id: "85c8dcb0-7af4-4581-b684-32087d386384",
            status: Enabled,
            cost: 0,
            condition: AutomodMessageHoldV2 {
                broadcaster_user_id: "129546453",
                moderator_user_id: "129546453",
            },
            created_at: "2024-11-18T16:36:08.691979783Z",
            transport: Websocket(
                WebsocketTransportResponse {
                    session_id: "AgoQREw4FYBWQ5quz4J-S4VYkRIGY2VsbC1j",
                    connected_at: None,
                    disconnected_at: None,
                },
            ),
            type_: AutomodMessageHold,
            version: "2",
        },
        message: Notification(
            AutomodMessageHoldV2Payload {
                broadcaster_user_id: "129546453",
                broadcaster_user_login: "nerixyz",
                broadcaster_user_name: "nerixyz",
                user_id: "489584266",
                user_login: "uint128",
                user_name: "uint128",
                message_id: "78ccd959-3e7e-4f8d-bd8b-f92c359b0a7d",
                message: AutomodMessage {
                    text: "ðŸ˜‚ ass",
                    fragments: [
                        Text {
                            text: "ðŸ˜‚ ",
                        },
                        Text {
                            text: "ass",
                        },
                    ],
                },
                reason: Automod(
                    AutomodMessageInfo {
                        category: Swearing,
                        level: 4,
                        boundaries: [
                            AutomodMessageBoundary {
                                start_pos: 2,
                                end_pos: 4,
                            },
                        ],
                    },
                ),
                held_at: "2024-11-18T16:59:46.323937273Z",
            },
        ),
    },
)
